aea/agent_loop.py

```diff
@@ -0,0 +1,184 @@
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+"""This module contains the implementation of an agent loop using asyncio."""
+
+import asyncio
+import time
+from asyncio.exceptions import CancelledError
+from asyncio.futures import Future
+from asyncio.tasks import create_task
+
+
+def ensure_list(value):
+    if not isinstance(value, list):
+        value = [value]
+    return value
+
+
+class AsyncState:
+    def __init__(self, initial_state=None, loop=None):
+        self._state = initial_state
+        self._watchers = set()
+        self._loop = loop or asyncio.get_event_loop()
+
+    @property
+    def state(self):
+        return self._state
+
+    def set(self, state):
+        if self._state == state:
+            return
+        self._state_changed(state)
+        self._state = state
+
+    def _state_changed(self, state):
+        for watcher in list(self._watchers):
+            if state in watcher._states:
+                self._loop.call_soon_threadsafe(
+                    watcher.set_result, (self._state, state)
+                )
+
+    async def wait(self, state_or_states):
+        states = ensure_list(state_or_states)
+
+        if self._state in states:
+            return (None, self._state)
+
+        watcher = Future()
+        watcher._states = states
+        self._watchers.add(watcher)
+        try:
+            return await watcher
+        finally:
+            self._watchers.remove(watcher)
+
+
+async def cancel_tasks_and_wait(tasks):
+    for task in tasks:
+        if not task.done():
+            task.cancel()
+        try:
+            await task
+        except CancelledError:
+            pass
+
+
+class ScheduledBehaviour:
+    def __init__(self, behaviour, loop=None):
+        self._loop = loop or asyncio.get_event_loop()
+        self.behaviour = behaviour
+        self._timerhandle = None
+
+    def _callback(self):
+        self._schedule_call()
+        self._call_act()
+
+    def _call_act(self):
+        self.behaviour.act()
+
+    def _schedule_call(self):
+        if self.behaviour._tick_interval is None:
+            return
+
+        if self._timerhandle is None:
+            ts = time.mktime(self.behaviour._start_at.timetuple())
+            delay = max(0, ts - time.time())
+            self._timerhandle = self._loop.call_later(delay, self._callback)
+        else:
+            self._timerhandle = self._loop.call_later(
+                self.behaviour._tick_interval, self._callback
+            )
+
+    def start(self):
+        if self._timerhandle:
+            return
+        self._schedule_call()
+
+    def stop(self):
+        if not self._timerhandle:
+            return
```

Comment:
It would be more relevant to see this in the context of the existing behaviour class.